### PR TITLE
Add better support for es6 modules.

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <gordonjsmith@gmail.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
FYI - The main issue that hit me, was that simply including d3-tip into my project forced the entire d3 library to be included (adding 600k to my project).

This code uses the typescript compiler to convert any es6 code to es5 (babel could have been an alternative).

I am not happy with the iife name "d3-tip", but that can be configured in the webpack config (I think its the only breaking change on what was there previously).

File: bower.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
     "docs"
   ],
   "dependencies": {
+    "d3": "^4.0.0",
```
Comment (GordonSmith) [inline bower.json]: Only needed by bar.html demo

File: examples/bars.html [hunk 2]
```diff
@@ -51,7 +51,7 @@
 
   </div>
   <script type="text/javascript">
-    var tip = d3.tip()
+    var tip = window["d3-tip"].tip()  //  Not sure this is an acceptable approach...
```
Comment (GordonSmith) [inline examples/bars.html]: I am not 100% ok with this - but am not sure what the best d3 v4 approach is.

**Comment by vdh** [comment]
There's already been an update to d3 v4 modules (`d3-collection` & `d3-selection`) in the latest master branch, it just hasn't been put into a release yet.

**Comment by GordonSmith** [comment]
@vdh Yes - I did notice that _after_ I added all the build stuff!  

There are still a few fixes in there, which you may want to throw an eye at - and the build process _may_ be of interest, but feel free to close this PR if you wish.

If you like I could convert the entire project to TypeScript, which tends to be an interesting exercise as typically it will show a bunch of potential issues (even if your not interested in pulling it)!

Like I say I won't be offended if you wish to close...

**Comment by vdh** [comment]
@GordonSmith I'm just a collaborator, not the project owner, and unfortunately not one with a lot of free time.

I'm a big fan of ES6, but it's more important to get unit testing added first (#165), so that any conversions or additional features don't break any existing functionality. There's already been some `rootElement`-related issues cropping up trying to merge in updates from other contributors into an alpha build.

**Comment by Caged** [comment]
The latest release now uses modern import/export constructs.

**Pull request closed without merging**